### ISSUE TITLE
Feature Request: Add new validation function called "IsUnique"

### ISSUE THREAD

**Issue by DBailey635** [Orckestra/CMS-Foundation#308]
This is a request for a new datatype field validation widget at Composite.Utils.Validation.IsUnique, that would check an entered value against other stored values for that field, checking if the entered value is unique.

This would allow the creation of asset registers, user directories, etc. It could even allow making any field the datatype key.

**Comment by burningice2866** [comment]
@DBailey635 You can define what field should be the key via the `KeyPropertyName` attribute like this

```
[KeyPropertyName("MyKeyField")]
public interface MyDataType : IData
{
   string MyKeyField { get; set; }
}
```

And in another type you can have another key

```
[KeyPropertyName("Name")]
public interface MyOtherDataType : IData
{
   string Name { get; set; }
   string Address { get; set; }
}
```

**Comment by DBailey635** [comment]
Thanks @burningice2866. I think I understand that.

There are occasions where I'd want both an Id field as the primary key, and one or more additional separate unique field values (the asset register example is one that I've worked on in the past), so my request for a new validation widget still stands. It'll make it easier to add secondary keys in the console.

**Comment by burningice2866** [comment]
@DBailey635 in such cases you can create a compound key, that is that your key is made up of 2 or more fields.

```
[KeyPropertyName("Name")]
[KeyPropertyName("Address ")]
public interface MyOtherDataType : IData
{
   string Name { get; set; }
   string Address { get; set; }
}
```

In such case, its the combination of Name and Address that is required to be unique, ie. you can have multiple persons named Trump, but only one Trump living at each address.

**Comment by burningice2866** [comment]
@DBailey635 i think i maybe understand your question regarding keys now... what I've been explaining is Primary Keys. Unfortunately C1 only supports Primary Keys - which is therefor just named **Key**.

If you're referring to, that you in ie. SQL Server can have a primary key and a normal key, then that is not supported by C1's datalayer. If you are running SQL Server, then you can do it manually on your table, but if you're on XML Datastore, the only option, as you started by asking for, is to create a validator which checks for uniqueness every time you add or update. 

That's unfortunately very inefficient since you're basically doing what is referred to as a Table Scan, instead of a Index Scan. every time you add or update - that is, going through all your existing records one by one, to see if there is a match.

**Comment by DBailey635** [comment]
I don't think it would be that inefficient. Wouldn't that be the equivalent of `SELECT COUNT(column_name) FROM table_name WHERE field = 'entered_value'` then checking if the Count is greater than zero? It would only occur upon saving a record.

**Comment by burningice2866** [comment]
Unless you have an Index on 'field', then it certainly is inefficient.

No Index = Table Scan = looping through each and every record.

**Comment by DBailey635** [comment]
Could adding the IsUnique widget trigger the addition of an index? Is that possible in .NET I the way this CMS works?

**Comment by burningice2866** [comment]
I just remembered that there is a [IndexAttribute](https://github.com/Orckestra/CMS-Foundation/blob/dev/Composite/Data/IndexAttribute.cs) that can be applied to ones fields if you want them to be indexed.

```
[KeyPropertyName("Id")] 
[Index("FolderName", IndexDirection.Ascending, "FileName", IndexDirection.Descending)]
interface IMyDataType : IData
{
    Guid Id { get; set; }
    string FolderName { get; set; }
    string FileName { get; set; }

    // other data type properties
}
```

**Comment by burningice2866** [comment]
But remember, that index doesn't guarantee its uniqueness

**Comment by burningice2866** [comment]
So whats missing, as i see it, is a way to define both Primary keys and Unique keys... only primary keys are currently supported by C1

**Comment by burningice2866** [comment]
And back to your original question... yes, its possible then to create a IsUnique validator for a field, and as long as we remember to also Index the field, it should be okay. Still, it's not handled by Datastore (ie SQL), but in the C1 Application layer, so if you happen to be running on SQL, i would recommend to manually add a Unique Key to your column in your database.

**Comment by burningice2866** [comment]
@DBailey635 what version of the CMS are you using? I could make a validator for you you can use.

**Comment by DBailey635** [comment]
Thanks @burningice2866. That's very generous. I'm using v5.2 for my current project, but it would be great if you could submit this as a pull request to the main project as well. 👍
